add doc for record OOM alarm (#11577)
pingcap/docs-cn: configure-memory-usage.md

## tidb-server 内存占用过高时的报警

默认配置下，tidb-server 实例会在机器内存使用达到总内存量的 70% 时打印报警日志，并记录相关状态文件。该内存使用率可以通过配置系统变量 [`tidb_memory_usage_alarm_ratio`](/system-variables.md#tidb_memory_usage_alarm_ratio) 进行设置。具体报警规则请参考该配置项的说明部分。

注意，内存使用量超过阈值，且满足如下任一条件时，才会进行告警，1. 第一次触发内存阈值；2. 触发内存阈值，且距离上一次告警超过 60s；3. 触发内存阈值，且(本次内存用量-上次告警时内存用量)/ 最大可用内存量 > 10%。

此外，为避免报警时产生的状态文件积累过多，目前默认保留最近 5 次报警时所生成的状态文件，保留文件数量可通过系统变量 [`tidb_memory_usage_alarm_keep_record_num`](/system-variables.md#tidb_memory_usage_alarm_keep_record_num) 配置。

下例通过构造一个占用大量内存的 SQL 语句触发报警，对该报警功能进行演示：

> 1. 配置报警比例为 `0.7`：

Author: > tidb_memory_usage_alarm_ratio 默认值就为 0.7，这里如果换一个值举例，比如 0.8 是否更清楚呢

change to 0.85

## tidb-server 内存占用过高时的报警

> 默认配置下，tidb-server 实例会在机器内存使用达到总内存量的 70% 时打印报警日志，并记录相关状态文件。该内存使用率可以通过配置系统变量 [`tidb_memory_usage_alarm_ratio`](/system-variables.md#tidb_memory_usage_alarm_ratio) 进行设置。具体报警规则请参考该变量的文档。

Reviewer: +1

## tidb-server 内存占用过高时的报警

> 当 tidb-server 实例的内存使用量超过内存阈值（默认为总内存量的 70%）且满足以下任一条件时，TiDB 将打印报警日志，并记录相关状态文件。

Reviewer 2: ```suggestion
当 tidb-server 实例的内存使用量超过内存阈值（默认为总内存量的 70%）且满足以下任一条件时，TiDB 将记录相关状态文件，并打印报警日志。
```

Reviewer 2: reason: 看起来是先记录状态文件，再打印报警日志

- 第一次内存使用量超过内存阈值。
- 内存使用量超过内存阈值，且距离上一次报警超过 60s。

> - 内存使用量超过内存阈值，且 `(本次内存使用量 - 上次报警时内存使用量) / 最大可用内存量 > 10%`。

Author: > 这里的 “最大可用内存量” 就是 L63 提到的总内存量吗？

是的

> - 内存使用量超过内存阈值，且 `(本次内存使用量 - 上次报警时内存使用量) / 最大可用内存量 > 10%`。

Reviewer 2: 那可以统一下措施吗，你看统一用哪个好呢？

你可以通过系统变量 [`tidb_memory_usage_alarm_ratio`](/system-variables.md#tidb_memory_usage_alarm_ratio) 修改触发该报警的内存使用比率，从而控制内存报警的阈值。

> 当打印报警日志时，TiDB 会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句的相关信息、goroutine 栈信息以及堆内存使用状态记录到 TiDB 日志文件 [`filename`](tidb-configuration-file.md#filename) 所在目录中，并输出一条包含关键字 `tidb-server has the risk of OOM` 以及以下内存相关系统变量的日志。

Reviewer 2: ```suggestion
当触发 tidb-server 内存占用过高的报警时，TiDB 的报警行为如下：

- TiDB 将以下信息记录到 TiDB 日志文件 [`filename`](tidb-configuration-file.md#filename) 所在目录中。

    - 当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句的相关信息
    - goroutine 栈信息
    - 堆内存使用状态

- TiDB 将输出一条包含关键字 `tidb-server has the risk of OOM` 以及以下内存相关系统变量的日志。
```

> - [`tidb_mem_oom_action`](#tidb_mem_oom_action-span-classversion-mark-v610-span)
> - [`tidb_mem_quota_query`](#tidb_mem_quota_query)
> - [`tidb_server_memory_limit`](#tidb_server_memory_limit)
> - [`tidb_analyze_version`](#tidb_analyze_version-span-classversion-mark-v510-span)
> - [`tidb_enable_rate_limit_action`](#tidb_enable_rate_limit_action)

Reviewer 2: ```suggestion
    - [`tidb_mem_oom_action`](#tidb_mem_oom_action-span-classversion-mark-v610-span)
    - [`tidb_mem_quota_query`](#tidb_mem_quota_query)
    - [`tidb_server_memory_limit`](#tidb_server_memory_limit)
    - [`tidb_analyze_version`](#tidb_analyze_version-span-classversion-mark-v510-span)
    - [`tidb_enable_rate_limit_action`](#tidb_enable_rate_limit_action)
```

## tidb-server 内存占用过高时的报警

当 tidb-server 实例的内存使用量超过内存阈值（默认为总内存量的 70%）且满足以下任一条件时，TiDB 将记录相关状态文件，并打印报警日志。

- 第一次内存使用量超过内存阈值。
- 内存使用量超过内存阈值，且距离上一次报警超过 60 秒。
- 内存使用量超过内存阈值，且 `(本次内存使用量 - 上次报警时内存使用量) / 总内存量 > 10%`。

你可以通过系统变量 [`tidb_memory_usage_alarm_ratio`](/system-variables.md#tidb_memory_usage_alarm_ratio) 修改触发该报警的内存使用比率，从而控制内存报警的阈值。

当触发 tidb-server 内存占用过高的报警时，TiDB 的报警行为如下：

> - TiDB 将以下信息记录到 TiDB 日志文件 [`filename`](tidb-configuration-file.md#filename) 所在目录中。

Reviewer 2: ```suggestion
- TiDB 将以下信息记录到 TiDB 日志文件 [`filename`](/tidb-configuration-file.md#filename) 所在目录中。
```

    - 当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句的相关信息
    - goroutine 栈信息
    - 堆内存使用状态

- TiDB 将输出一条包含关键字 `tidb-server has the risk of OOM` 以及以下内存相关系统变量的日志。

    - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-从-v610-版本开始引入)
    - [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)

>     - [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit)

Reviewer 2: ```suggestion
    - [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入)
```

>     - [`tidb_mem_oom_action`](#tidb_mem_oom_action-span-classversion-mark-v610-span)
>     - [`tidb_mem_quota_query`](#tidb_mem_quota_query)
>     - [`tidb_server_memory_limit`](#tidb_server_memory_limit)
>     - [`tidb_analyze_version`](#tidb_analyze_version-span-classversion-mark-v510-span)
>     - [`tidb_enable_rate_limit_action`](#tidb_enable_rate_limit_action)

Reviewer 2: ```suggestion
    - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-span-classversion-mark-v610-span)
    - [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)
    - [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit)
    - [`tidb_analyze_version`](/system-variables.md#tidb_analyze_version-span-classversion-mark-v510-span)
    - [`tidb_enable_rate_limit_action`](/system-variables.md#tidb_enable_rate_limit_action)
```

>     - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-span-classversion-mark-v610-span)
>     - [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)
>     - [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit)
>     - [`tidb_analyze_version`](/system-variables.md#tidb_analyze_version-span-classversion-mark-v510-span)

Reviewer 2: ```suggestion
    - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-从-v610-版本开始引入)
    - [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)
    - [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit)
    - [`tidb_analyze_version`](/system-variables.md#tidb_analyze_version-从-v510-版本开始引入)
```

> 为避免报警时产生的状态文件累积过多，目前 TiDB 默认只保留最近 5 次报警时所生成的状态文件。你可以通过配置系统变量 [`tidb_memory_usage_alarm_keep_record_num`](/system-variables.md#tidb_memory_usage_alarm_keep_record_num) 调整该次数。

Reviewer 2: ```suggestion
为避免报警时产生的状态文件累积过多，目前 TiDB 默认只保留最近 5 次报警时所生成的状态文件。你可以通过配置系统变量 [`tidb_memory_usage_alarm_keep_record_num`](/system-variables.md#tidb_memory_usage_alarm_keep_record_num-从-v640-版本开始引入) 调整该次数。
```
